tests/src/Simple/Threading/Threading.cs

```diff
@@ -758,13 +793,172 @@ public static void MutexMaximumReacquireCountTest()
+    }
+
+    [Fact]
+    public static void CanRegisterMoreThan64Waits()
+    {
+        RegisteredWaitHandle[] handles = new RegisteredWaitHandle[65];
+        for(int i = 0; i < 65; ++i) {
+            handles[i] = ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(false), (_, __) => {}, null, -1, true);
+        }
+        for(int i = 0; i < 65; ++i) {
+            handles[i].Unregister(null);
+        }
+    }
+
+    [Fact]
+    public static void StateIsPasssedThroughToCallback()
+    {
+        object state = new object();
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(true), (callbackState, _) =>
+        {
+            if(state == callbackState)
+            {
+                e0.Set();
+            }
+        }, state, 0, true);
+        Assert.True(e0.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void WaitWithLongerTimeoutThanWaitThreadCanStillTimeout()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(false), (_, __) => e0.Set(), null, 21000, true);
+        Thread.Sleep(20000);
+        Assert.True(e0.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void UnregisterCallbackIsNotCalledAfterCallbackFinishesIfAnotherCallbackOnSameWaitRunning()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        AutoResetEvent e1 = new AutoResetEvent(false);
+        RegisteredWaitHandle handle = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) =>
+        {
+            Thread.Sleep(300);
+        }, null, ThreadTestHelpers.UnexpectedTimeoutMilliseconds, false);
+        e0.Set();
+        Thread.Sleep(50);
+        e0.Set();
+        Thread.Sleep(50);
+        handle.Unregister(e1);
+        Assert.False(e1.WaitOne(ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void CallingUnregisterOnAutomaticallyUnregisteredHandleReturnsTrue()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        RegisteredWaitHandle handle = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) => {}, null, ThreadTestHelpers.UnexpectedTimeoutMilliseconds, true);
+        e0.Set();
+        Thread.Sleep(ThreadTestHelpers.ExpectedTimeoutMilliseconds);
+        Assert.True(handle.Unregister(null));
+    }
+
+    [Fact]
+    public static void EventSetAfterUnregisterNotObservedOnWaitThread()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        RegisteredWaitHandle handle = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) => {}, null, ThreadTestHelpers.UnexpectedTimeoutMilliseconds, true);
+        handle.Unregister(null);
+        e0.Set();
+        Assert.True(e0.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+}
+
```

Comment:
These tests are pretty good, thanks for adding them. We have previously talked about some scenarios where in some scenario something needed fixing, it would be nice to review if there are any more of those and add tests for them as well. For issues involving races, it may not be possible to cause the order of events to happen exactly as desired but it may still be a good test especially when used as part of a stress run.